feat(AIP-203): introduce IDENTIFIER field behavior

toumorokoshi opened the pull request:
Identifier fields are often multi-modal: they are required or optional
in various situations. Adding an explicit field behavior for these
fields helps eliminate ambiguity for an author to fully annotate field
behavior on all request fields, as well as clarify which fields to use
in identifying the resource for clients.

fixes #1176.

toumorokoshi committed:
chore(AIP-203): resorting field behavior

Making it a bit easier to find the appropriate field behavior with
alphabetical sorting.

toumorokoshi committed:
feat(AIP-203): introduce IDENTIFIER field behavior

Identifier fields are often multi-modal: they are required or optional
in various situations. Adding an explicit field behavior for these
fields helps eliminate ambiguity for an author to fully annotate field
behavior on all request fields, as well as clarify which fields to use
in identifying the resource for clients.

fixes #1176.

noahdietz commented on aip/general/0203.md:
```diff
@@ -44,42 +44,51 @@ document this behavior for clients.
-The use of `REQUIRED` indicates that the field **must** be present (and set to
-a non-empty value) on the request or resource.
+The use of `IDENTIFIER` indicates that the field is used to identify the
+resource. It must be attached to the name parameter and any other fields that
```
```suggestion
resource. It **must** be attached to the name parameter and any other fields that
```

noahdietz commented on aip/general/0203.md:
```diff
@@ -44,42 +44,51 @@ document this behavior for clients.
+The use of `IDENTIFIER` indicates that the field is used to identify the
+resource. It must be attached to the name parameter and any other fields that
+are used in identifying the resource within the collection. See [fields
+representing resource names].
```
Is this to say we want to see it used outside of messages that represent resources? If so, couldn't that conflict with other field_behavior annotations?

The sentence following kind of contradicts the second part of this first sentence...

toumorokoshi replied:
Are you referring to the reference to "fields representing resource names"? That's just the name of the section, although really that section could be called "resource names within resources" or something like that.

I linked to the section to try to ensure we don't have redundant (and therefore hard to maintain) guidance.

I tried to reword the section altogether to call out fields within a resource used to identify the resource. PTAL.

noahdietz replied:
Apologies, GitHub doesn't let me highlight a specific section for comment. It was this section that threw me off:
> It must be attached to...and any other fields within a resource that are used in identifying the resource within a collection

What resource fields, outside of `name`, would need the annotation? I worry that a `string book_id = 2 OUTPUT_ONLY` will also get `IDENTIFIER` added. Or is that the intention?

The new language does help though!

toumorokoshi replied:
I was originally thinking we may want to account for some older services like compute, but based on what I'm seeing, they don't really have field identifiers.

even older services like bigquery still have name: https://github.com/googleapis/googleapis/blob/d994e780c732da0a902070d0314ca401fb6132df/google/cloud/bigquery/reservation/v1/reservation.proto#L1120-L1121.

So I'll just update the wording that it should be exclusively for name, as that's the longterm guidance anyway.

noahdietz replied:
Thanks, I see what you are saying now and I agree with your analysis.

toumorokoshi replied:
ok updated, PTAL.

noahdietz replied:
Yep, looks good, thanks!

noahdietz commented on aip/general/0203.md:
```diff
@@ -44,42 +44,51 @@ document this behavior for clients.
+representing resource names].
 
-A field **should** only be described as required if _either_:
+This annotation should not be applied to references to other resources within a
```
```suggestion
This annotation **should not** be applied to references to other resources within a
```

noahdietz reviewed (changes requested):
Can you include the Compatibility guidance in here as well (you documented it in the proposal doc).

Also, can you update other AIPs that portray the `google.api.resource` with a `name` field to include the IDENTIFIER value?

toumorokoshi committed:
accept suggestions

andrei-scripniciuc commented on aip/general/0203.md:
```diff
@@ -44,42 +44,51 @@ document this behavior for clients.
-- A field which is conditionally required in some situations.
-- A field on any message (including messages that are resources) which is never
-  used as user input.
+- Names or IDs which are set on creation and then used as a primary key.
```
perhaps I'm reading this wrong, but are we saying that IDENTIFIER fields could also be marked as IMMUTABLE? names shouldn't be mutable but the IDENTIFIER field should be enough to infer that.

toumorokoshi replied:
ack, yeah this should be updated. removed the identifier use case, replaced it with something more practical.

Yes, agreed: you should only need `IDENTIFIER`.

andrei-scripniciuc commented on aip/general/0203.md:
```diff
@@ -112,41 +121,42 @@ are:
-Potential use cases for immutable fields (this is not an exhaustive list) are:
+Fields **should not** be described as required in order to signify:
 
-- Names or IDs which are set on creation and then used as a primary key.
```
same comment here, shouldn't we call out that name / id fields should not be marked as required, since the IDENTIFIER annotation will express that?

toumorokoshi replied:
fixed PTAL.

toumorokoshi committed:
address comments

toumorokoshi committed:
Merge branch 'master' into toum/203-identifier

toumorokoshi committed:
addressing feedback

- updating name examples with the field behavior
- adding backwards compatibility considerations

noahdietz commented on aip/general/0203.md:
```diff
@@ -44,6 +44,27 @@ document this behavior for clients.
+This annotation **must not** be applied to references to other resources within
+a message.
+
+#### Identifier Backwards Compatibility
```
We already have a `Backwards Compatibility` section for all interesting cases. Should this stuff go there?

toumorokoshi replied:
good call, will move.

toumorokoshi committed:
Merge branch 'master' into toum/203-identifier
